File: vortex-sdk-python/vortex_sdk/exceptions.py
"""
Custom exceptions for the Vortex SDK.
"""

class VortexException(Exception):
    """Base exception for all Vortex SDK errors."""
    pass

class VortexConnectionError(VortexException):
    """Raised when there's an issue connecting to the Vortex server."""
    pass

class VortexTimeoutError(VortexException):
    """Raised when an operation times out."""
    pass

class VortexApiError(VortexException):
    """Raised for errors returned by the Vortex API."""
    def __init__(self, message: str, grpc_error: Exception = None, status_code: int = None, details: str = None):
        super().__init__(message)
        self.grpc_error = grpc_error
        self.status_code = status_code
        self.details = details

        if grpc_error:
            if hasattr(grpc_error, 'code') and callable(grpc_error.code):
                try:
                    # grpc.StatusCode is an enum, we might want its name or value
                    grpc_status_code = grpc_error.code()
                    if hasattr(grpc_status_code, 'name'):
                        self.status_code = grpc_status_code.name 
                    elif hasattr(grpc_status_code, 'value') and isinstance(grpc_status_code.value, tuple):
                         self.status_code = grpc_status_code.value[0] # value can be (int, str)
                    else:
                        self.status_code = str(grpc_status_code)
                except Exception:
                    pass # Keep original status_code if any
            
            if hasattr(grpc_error, 'details') and callable(grpc_error.details):
                try:
                    self.details = grpc_error.details()
                except Exception:
                    pass # Keep original details if any
            elif not self.details: # If no details method, use string representation of error
                self.details = str(grpc_error)


    def __str__(self):
        base_str = super().__str__()
        if self.status_code:
            base_str += f" (Status Code: {self.status_code})"
        if self.details:
            base_str += f" Details: {self.details}"
        return base_str

class VortexClientConfigurationError(VortexException):
    """Raised for client configuration errors."""
    pass

File: vortex-sdk-typescript/src/exceptions.ts
/**
 * Custom exceptions for the Vortex TypeScript SDK.
 */
import * as grpc from '@grpc/grpc-js'; // Import the grpc namespace
import { ServiceError } from '@grpc/grpc-js';

/**
 * Base class for all Vortex SDK specific exceptions.
 */
export class VortexException extends Error {
  /**
   * Creates an instance of VortexException.
   * @param message - The error message.
   */
  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;
  }
}

/**
 * Represents an error that occurs during connection to the Vortex server.
 */
export class VortexConnectionError extends VortexException {}

/**
 * Represents an error that occurs due to a timeout.
 */
export class VortexTimeoutError extends VortexException {}

/**
 * Represents an error returned by the Vortex API.
 * This class may wrap an underlying gRPC service error.
 */
export class VortexApiError extends VortexException {
  /**
   * The underlying gRPC ServiceError, if the API error originated from a gRPC call.
   */
  public readonly grpcError?: ServiceError;
  /**
   * The gRPC status code or a custom string status code associated with the API error.
   */
  public readonly statusCode?: grpc.status | string;
  /**
   * Additional details associated with the API error.
   */
  public readonly details?: string;
  /**
   * Flag indicating if this error was due to a client-side timeout.
   */
  public readonly isClientTimeout: boolean;

  /**
   * Creates an instance of VortexApiError.
   * @param message - The primary error message.
   * @param options - Optional parameters to include gRPC error details.
   * @param options.grpcError - The original gRPC ServiceError.
   * @param options.statusCode - A specific status code (gRPC or custom). Defaults to `grpcError.code`.
   * @param options.details - Additional error details. Defaults to `grpcError.details`.
   * @param options.isClientTimeout - Whether this error was due to a client-side timeout.
   */
  constructor(message: string, options?: {
    grpcError?: ServiceError;
    statusCode?: grpc.status | string;
    details?: string;
    isClientTimeout?: boolean;
  }) {
    super(message);
    this.grpcError = options?.grpcError;
    this.statusCode = options?.statusCode || options?.grpcError?.code;
    this.details = options?.details || options?.grpcError?.details;
    this.isClientTimeout = options?.isClientTimeout || false;
  }

  /**
   * Returns a string representation of the API error, including status code and details if available.
   * @returns A string describing the error.
   */
  public toString(): string {
    let str = super.toString();
    if (this.statusCode !== undefined) {
      str += ` (Status Code: ${this.statusCode})`;
    }
    if (this.details) {
      str += ` Details: ${this.details}`;
    }
    return str;
  }
}

/**
 * Represents an error related to client configuration.
 */
export class VortexClientConfigurationError extends VortexException {}
